File: backend/app/schemas/asignacion_servicio.py
from datetime import datetime
from typing import Optional
from pydantic import BaseModel

from app.models.asignacion_servicio import EstadoAsignacion


# ── Creación (admin asigna mecánico a un incidente) 
class AsignacionCreate(BaseModel):
    incidente_id: int
    mecanico_id: Optional[int] = None   # opcional: el admin puede aceptar sin asignar mecánico aún
    costo_estimado: Optional[float] = None
    distancia_km: Optional[float] = None
    tiempo_estimado: Optional[int] = None


# ── Asignar mecánico a una asignación ya aceptada
class AsignacionAsignarMecanico(BaseModel):
    mecanico_id: int


# ── Rechazo (mecánico rechaza con motivo)
class AsignacionRechazar(BaseModel):
    motivo_rechazo: str


# ── Actualización de estado (mecánico avanza el estado) 
class AsignacionEstadoUpdate(BaseModel):
    estado: EstadoAsignacion


# ── Resúmenes anidados (para que el front identifique la asignación)
class IncidenteResumen(BaseModel):
    id: int
    descripcion: str
    fecha_hora: datetime
    latitud: Optional[float] = None    # ubicación del cliente (para el mapa del mecánico)
    longitud: Optional[float] = None

    model_config = {"from_attributes": True}


class UsuarioResumen(BaseModel):
    nombre: str
    apellido: str

    model_config = {"from_attributes": True}


class MecanicoResumen(BaseModel):
    id: int
    usuario: Optional[UsuarioResumen] = None

    model_config = {"from_attributes": True}


# ── Lectura
class AsignacionRead(BaseModel):
    id:              int
    incidente_id:    int
    mecanico_id:     Optional[int]
    costo_estimado:  Optional[float]
    distancia_km:    Optional[float]
    tiempo_estimado: Optional[int]
    estado:          EstadoAsignacion
    fecha_creacion:  datetime
    fecha_respuesta: Optional[datetime]
    motivo_rechazo:  Optional[str]
    incidente:       Optional[IncidenteResumen] = None
    mecanico:        Optional[MecanicoResumen] = None

    model_config = {"from_attributes": True}

File: movil/lib/models/asignacion_servicio.dart
import 'package:equatable/equatable.dart';
import 'package:json_annotation/json_annotation.dart';

part 'asignacion_servicio.g.dart';

enum EstadoAsignacion {
  @JsonValue('pendiente')
  pendiente,
  @JsonValue('buscando_taller')
  buscandoTaller,
  @JsonValue('taller_asignado')
  tallerAsignado,      
  @JsonValue('en_camino')
  enCamino,
  @JsonValue('en_atencion')
  enAtencion,          
  @JsonValue('finalizado')
  finalizado,          // ← CORREGIDO: era "completada"
  @JsonValue('cancelado')
  cancelado,           // ← CORREGIDO: era "cancelada"
  @JsonValue('rechazada')
  rechazada,
}

/// Resumen del incidente anidado en la asignación (lo envía el backend
/// para que el mecánico vea descripción, fecha y ubicación del cliente).
@JsonSerializable()
class IncidenteResumen extends Equatable {
  final int id;
  final String descripcion;
  @JsonKey(name: 'fecha_hora')
  final DateTime fechaHora;
  final double? latitud;
  final double? longitud;

  const IncidenteResumen({
    required this.id,
    required this.descripcion,
    required this.fechaHora,
    this.latitud,
    this.longitud,
  });

  factory IncidenteResumen.fromJson(Map<String, dynamic> json) =>
      _$IncidenteResumenFromJson(json);

  Map<String, dynamic> toJson() => _$IncidenteResumenToJson(this);

  @override
  List<Object?> get props => [id, descripcion];
}

@JsonSerializable()
class AsignacionServicio extends Equatable {
  final int id;
  @JsonKey(name: 'costo_estimado')
  final double? costoEstimado;
  @JsonKey(name: 'distancia_km')
  final double? distanciaKm;
  @JsonKey(name: 'tiempo_estimado')
  final int? tiempoEstimado;
  final EstadoAsignacion estado;
  @JsonKey(name: 'fecha_creacion')
  final DateTime fechaCreacion;
  @JsonKey(name: 'fecha_respuesta')
  final DateTime? fechaRespuesta;
  @JsonKey(name: 'motivo_rechazo')
  final String? motivoRechazo;
  @JsonKey(name: 'incidente_id')
  final int incidenteId;
  @JsonKey(name: 'mecanico_id')
  final int? mecanicoId;
  final IncidenteResumen? incidente;

  const AsignacionServicio({
    required this.id,
    this.costoEstimado,
    this.distanciaKm,
    this.tiempoEstimado,
    required this.estado,
    required this.fechaCreacion,
    this.fechaRespuesta,
    this.motivoRechazo,
    required this.incidenteId,
    this.mecanicoId,
    this.incidente,
  });

  factory AsignacionServicio.fromJson(Map<String, dynamic> json) =>
      _$AsignacionServicioFromJson(json);

  Map<String, dynamic> toJson() => _$AsignacionServicioToJson(this);

  @override
  List<Object?> get props => [id, estado, incidenteId, mecanicoId];
}